telegraf/plugins/plugin.go

```diff
@@ -0,0 +1,1533 @@
+package plugins
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+// Plugin defines a Telegraf plugin.
+type Plugin struct {
+	Type        string `json:"type,omitempty"`        // Type of the plugin.
+	Name        string `json:"name,omitempty"`        // Name of the plugin.
+	Description string `json:"description,omitempty"` // Description of the plugin.
+	Config      string `json:"config,omitempty"`      // Config contains the toml config of the plugin.
+}
+
+// TelegrafPlugins defines a Telegraf version's collection of plugins.
+type TelegrafPlugins struct {
+	Version string   `json:"version,omitempty"` // Version of telegraf plugins are for.
+	OS      string   `json:"os,omitempty"`      // OS the plugins apply to.
+	Plugins []Plugin `json:"plugins,omitempty"` // Plugins this version of telegraf supports.
+}
+
+// ListAvailablePlugins lists available plugins based on type.
+func ListAvailablePlugins(t string) (*TelegrafPlugins, error) {
+	switch t {
+	case "input":
+		return AvailableInputs()
+	case "output":
+		return AvailableOutputs()
+	case "processor":
+		return AvailableProcessors()
+	case "aggregator":
+		return AvailableAggregators()
+	default:
+		return nil, fmt.Errorf("unknown plugin type '%s'", t)
+	}
+}
+
+// AvailablePlugins returns the base list of available plugins.
+func AvailablePlugins() (*TelegrafPlugins, error) {
+	all := &TelegrafPlugins{}
+
+	t, err := AvailableInputs()
+	if err != nil {
+		return nil, err
+	}
+	all.Version = t.Version
+	all.Plugins = append(all.Plugins, t.Plugins...)
+
+	t, err = AvailableOutputs()
+	if err != nil {
+		return nil, err
+	}
+	all.Plugins = append(all.Plugins, t.Plugins...)
+
+	t, err = AvailableProcessors()
+	if err != nil {
+		return nil, err
+	}
+	all.Plugins = append(all.Plugins, t.Plugins...)
+
+	t, err = AvailableAggregators()
+	if err != nil {
+		return nil, err
+	}
+	all.Plugins = append(all.Plugins, t.Plugins...)
+
+	return all, nil
+}
+
+// AvailableInputs returns the base list of available input plugins.
+func AvailableInputs() (*TelegrafPlugins, error) {
```

Comment:
what's the motivation for using pointer semantics as the return type here? Feels like you're better off using value semantics for the TelegrafPlugins type 🤷‍♂, just curious